sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala

```diff
@@ -126,7 +126,15 @@ final class Decimal extends Ordered[Decimal] with Serializable {
   def set(decimal: BigDecimal): Decimal = {
     this.decimalVal = decimal
     this.longVal = 0L
-    this._precision = decimal.precision
+    if (decimal.compare(BigDecimal(1.0)) == -1 && decimal.compare(BigDecimal(-1.0)) == 1) {
```

Comment:
Don't we allow the case that precision equals to scale? Seems it works for the example expressions.

    scala> sql("select floor(0.9)").show()
    +----------+
    |FLOOR(0.9)|
    +----------+
    |         0|
    +----------+


    scala> sql("select ceil(0.9)").show()
    +---------+
    |CEIL(0.9)|
    +---------+
    |        1|
    +---------+


    scala> sql("select 1 > 0.9").show()
    +-----------------------+
    |(CAST(1 AS BIGINT) > 0)|
    +-----------------------+
    |                   true|
    +-----------------------+